pennylane_lightning_gpu/src/algorithms/ObservablesGPU.hpp

```diff
@@ -497,77 +497,79 @@ class SparseHamiltonianGPU final : public ObservableGPU<T> {
             /* cusparseIndexType_t */ compute_type,
             /* cusparseIndexBase_t */ CUSPARSE_INDEX_BASE_ZERO,
-            /* cudaDataType */ data_type))
+            /* cudaDataType */ data_type));
 
-        // Create dense vector X
+        // Create dense statevector descriptor
         PL_CUSPARSE_IS_SUCCESS(cusparseCreateDnVec(
-            /* cusparseDnVecDescr_t* */ &vecX,
+            /* cusparseDnVecDescr_t* */ &sv_descriptor,
             /* int64_t */ num_cols,
             /* void* */ sv.getData(),
-            /* cudaDataType */ data_type))
+            /* cudaDataType */ data_type));
 
-        // Create dense vector y
+        // Create dense statevector prime descriptor
         PL_CUSPARSE_IS_SUCCESS(cusparseCreateDnVec(
-            /* cusparseDnVecDescr_t* */ &vecY,
+            /* cusparseDnVecDescr_t* */ &sv_prime_descriptor,
             /* int64_t */ num_rows,
-            /* void* */ sv.getData(),
-            /* cudaDataType */ data_type))
+            /* void* */ d_sv_prime->getData(),
+            /* cudaDataType */ data_type));
 
         // allocate an external buffer if needed
         PL_CUSPARSE_IS_SUCCESS(cusparseSpMV_bufferSize(
             /* cusparseHandle_t */ handle,
             /* cusparseOperation_t */ CUSPARSE_OPERATION_NON_TRANSPOSE,
             /* const void* */ &alpha,
-            /* cusparseSpMatDescr_t */ mat,
-            /* cusparseDnVecDescr_t */ vecX,
+            /* cusparseSpMatDescr_t */ sparseH_descriptor,
+            /* cusparseDnVecDescr_t */ sv_descriptor,
             /* const void* */ &beta,
-            /* cusparseDnVecDescr_t */ vecY,
+            /* cusparseDnVecDescr_t */ sv_prime_descriptor,
             /* cudaDataType */ data_type,
             /* cusparseSpMVAlg_t */
             CUSPARSE_SPMV_CSR_ALG1, // Can also use CUSPARSE_MV_ALG_DEFAULT
-            /* size_t* */ &bufferSize))
+            /* size_t* */ &bufferSize));
 
-        DataBuffer<cudaDataType_t, int> dBuffer{bufferSize, device_id,
-                                                stream_id, true};
+        DataBuffer<cudaDataType_t> dBuffer{bufferSize, device_id, stream_id,
+                                           true};
 
         // execute SpMV
         PL_CUSPARSE_IS_SUCCESS(cusparseSpMV(
             /* cusparseHandle_t */ handle,
             /* cusparseOperation_t */ CUSPARSE_OPERATION_NON_TRANSPOSE,
             /* const void* */ &alpha,
-            /* cusparseSpMatDescr_t */ mat,
-            /* cusparseDnVecDescr_t */ vecX,
+            /* cusparseSpMatDescr_t */ sparseH_descriptor,
+            /* cusparseDnVecDescr_t */ sv_descriptor,
             /* const void* */ &beta,
-            /* cusparseDnVecDescr_t */ vecY,
+            /* cusparseDnVecDescr_t */ sv_prime_descriptor,
             /* cudaDataType */ data_type,
             /* cusparseSpMVAlg_t */
-            CUSPARSE_SPMV_CSR_ALG1, // Can also use CUSPARSE_MV_ALG_DEFAULT
+            CUSPARSE_MV_ALG_DEFAULT,
```

Comment:
Given that the CSR matrix will always be represented in CSR format, why change the algorithm to the default?
Is this necessary because of the stochastic nature of the `CUSPARSE_SPMV_CSR_ALG1`? 
If yes, did you try `CUSPARSE_SPMV_CSR_ALG2`, this one is slower but deterministic.

Reply:
Being specific is fine I guess, but I'd rather let the API decide.